library/old_protocol/serialize.cpp

```diff
@@ -0,0 +1,97 @@
+#include "serialize.hpp"
+
+#include <msgpack.hpp>
+
+#include "library/elliptics.h"
+
+namespace msgpack {
+
+using namespace ioremap::elliptics;
+
+template <typename Stream>
+inline msgpack::packer<Stream> &operator <<(msgpack::packer<Stream> &o, const dnet_time &v) {
+	o.pack_array(2);
+	o.pack_fix_uint64(v.tsec);
+	o.pack_fix_uint64(v.tnsec);
+	return o;
+}
+
+template <typename Stream>
+inline msgpack::packer<Stream> &operator<<(msgpack::packer<Stream> &o, const n2::lookup_response &v) {
+	o.pack_array(12);
+
+	o.pack(v.record_flags);
+	o.pack(v.user_flags);
+	o.pack(v.path);
+
+	o.pack(v.json_timestamp);
+	o.pack(v.json_offset);
+	o.pack(v.json_size);
+	o.pack(v.json_capacity);
+
+	o.pack(v.data_timestamp);
+	o.pack(v.data_offset);
+	o.pack(v.data_size);
+
+	o.pack(v.json_checksum);
+	o.pack(v.data_checksum);
+
+	return o;
+}
+
+} // namespace msgpack
+
+namespace ioremap { namespace elliptics { namespace n2 {
+
+void enqueue_net(dnet_net_state *st, net_iovec serialized) {
+	std::unique_ptr<n2_net_iovec> iov(new n2_net_iovec{ std::move(serialized) });
+
+	auto r = static_cast<dnet_io_req *>(calloc(1, sizeof(dnet_io_req)));
+	if (!r)
+		throw std::bad_alloc();
+
+	r->iov = iov.release();
+	n2_io_req_enqueue_net(st, r);
+}
+
+static uint64_t response_transform_flags(uint64_t flags) {
+	return (flags & ~(DNET_FLAGS_NEED_ACK)) | DNET_FLAGS_REPLY;
+}
+
+net_iovec serialize_error_response(const dnet_cmd &cmd_in) {
+	dnet_cmd cmd = cmd_in;
+	cmd.flags = response_transform_flags(cmd.flags);
+	dnet_convert_cmd(&cmd);
+
+	return { data_pointer::copy(&cmd, sizeof(dnet_cmd)) };
```

Comment:
Why is it copied twice?

Reply:
`data_pointer::copy` is used to dynamically allocate memory buffer for serialized response. Because every response must be serialized to `n2_iovec`.

We discussed that I should separate cmd out of message's tail. So I'll rework these serializators, and here won't be extra allocation (for messages that contain only dnet_cmd).

Reply:
Fixed: now I convert to data_pointer only message's tail.

Reply:
Reworked, now `cmd` is separated with buffer that is allocated only for message body